### PR TITLE
Add ax kwarg to matplotlib visualization functions

### LINKED ISSUE #1950: Visualizations should allow for passing a Matplotlib axes instance
### What is the expected enhancement?
Being able to pass an axes instance allows for easily using the visualzations in subplots and other types of figures where the figure instance is already generated

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds support for passing a matplotlib axes to visualization
functions. This enables integrating the qiskit visualizations into a
larger matplotlib visualization. When an ax kwarg is set that input will
be used for the output visualization and a new figure will not be
generated or returned.

### Details and comments

Fixes #1950

### CODE REVIEW

**Comment by 1ucian0** [comment]
Given that matplotlib calls the module `axes`, why not calling the argument `axes`?

File: releasenotes/notes/pass-mpl-ax-kwarg-069f793f01cd61a7.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+     * ``qiskit.visualization.plot_state_city``
+
+    These new kargs work the same as the newly added ``ax`` kwargs for other
+    visualization functions. However because these plots use two axes (one for
+    the real component, the other for the imaginary component). Having two
+    kwargs also provides the flexibility to only generate a visualization for
+    one of the components instead of always doing both. For example::
+
+        from matplotlib import pyplot as plt
+        from qiskit.visualization.plot_state_hinton
```
Comment (1ucian0) [inline releasenotes/notes/pass-mpl-ax-kwarg-069f793f01cd61a7.yaml]:
```suggestion
        from qiskit.visualization import plot_state_hinton
```

**Comment by mtreinish** [comment]
> It seems that the parameters are not optional. Previously working code is breaking :
> 
> ```python
> psi = [0.70710678118654746, 0, 0, 0.70710678118654746]
> plot_state_hinton(psi)
> ```
> 
> ```
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> <ipython-input-3-7e1883ce9c63> in <module>
>       1 psi = [0.70710678118654746, 0, 0, 0.70710678118654746]
> ----> 2 plot_state_hinton(psi)
> 
> ~/repos/qiskit-terra/qiskit/visualization/state_visualization.py in plot_state_hinton(rho, title, figsize, ax_real, ax_imag)
>      99             fig = ax_real.get_figure()
>     100         else:
> --> 101             fig = ax_imag.get_figure()
>     102         ax1 = ax_real
>     103         ax2 = ax_imag
> 
> AttributeError: 'NoneType' object has no attribute 'get_figure'
> ```
> 
> It's not obvious for me how this new options are used. Maybe a more explicit subplot example in the release notes?

It was a typo in `plot_state_hinton()` fixed in the latest patch. I'll take another look at release note and see if I can come up with a clearer example.

**Comment by 1ucian0** [comment]
do you have any thoughts about https://github.com/Qiskit/qiskit-terra/pull/3053#issuecomment-530112487 ?

**Comment by mtreinish** [comment]
> do you have any thoughts about [#3053 (comment)](https://github.com/Qiskit/qiskit-terra/pull/3053#issuecomment-530112487) ?

Sorry I missed that among the other review comments earlier. I really don't care about the name much, `ax` is also commonly used in the matplotlib documentation for things that deal with an `Axes` object. I'd be fine deferring to whatever your preference is, however in this case we are already using `ax` in an existing API, see: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/visualization/gate_map.py#L70-L80 so I used `ax` to be consistent.

File: .pylintrc [hunk 2]
```diff
@@ -172,7 +172,7 @@ attr-rgx=[a-z_][a-z0-9_]{2,30}$
 attr-name-hint=[a-z_][a-z0-9_]{2,30}$
 
 # Regular expression matching correct argument names
-argument-rgx=[a-z_][a-z0-9_]{2,30}$
+argument-rgx=[a-z_][a-z0-9_]{2,30}|ax$
```
Comment (1ucian0) [inline .pylintrc]: I think this allows to remove https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/visualization/gate_map.py#L14

**Comment by 1ucian0** [comment]
> however in this case we are already using ax in an existing API

Okey. That's a good argument to keep it. Let's merge.

File: .pylintrc (hunk 2 above)
Comment (mtreinish) [reply .pylintrc]: Yep, it does. Do you want me to do that here? Or do it after merging this?

Comment (1ucian0) [reply .pylintrc]: changed in 36613777

**Pull request merged**